### PR TITLE
Update finch to 0.12.1

### PR DESCRIPTION
**Pull request by tlogan2000**
## Overview
Update `finch` config to a more recent release : https://github.com/bird-house/finch/releases/tag/v0.12.1

## Changes

**Non-breaking changes**
- Adds `finch` v0.12.1

## Additional Information

Docstring issues coming from `finch.heat_wave_index` were found to be related to a bug/feature in `birdy` that causes parameter options presented in a set to be reported in an unordered way. The relevant issue can be found here: https://github.com/bird-house/birdy/issues/277

### CODE REVIEW

**Comment by tlogan2000** [comment]
@tlvu  I think this is all that's needed to deploy the new finch image?

**Comment by Zeitsperre** [comment]
@tlogan2000 not sure if it's of interest, but the dockerfile will be cleaned up in my PR. Is it worth waiting?

**Comment by tlogan2000** [comment]
> @tlogan2000 not sure if it's of interest, but the dockerfile will be cleaned up in my PR. Is it worth waiting?

yes no rush

**Review by fmigneault — CHANGES_REQUESTED** [review]
Needs fixes to the CI notebooks to support the new version.

**Comment by mishaschwartz** [comment]
@fmigneault 

> Needs fixes to the CI notebooks to support the new version.

https://github.com/Ouranosinc/pavics-sdi/pull/328

**Comment by Zeitsperre** [comment]
@tlvu I've fixed the merge issues in https://github.com/Ouranosinc/pavics-sdi/pull/328. Is there a way to run Jenkins CI against that branch based on this PR?

**Comment by tlvu** [comment]
> @tlvu My PAVICS host is completely offline. Guess we'll need to wait for CRIM infra to come back online.

Well ... learn to fix it :D  Because each time that you'll upgrade Finch or Raven or any WPS service, you'll need your PAVICS host to test.

CRIM pipeline can not really be used as a development workflow because the PAVICS instance is destroyed right after Jenkins completed so if you need to "just re-run" something to confirm a fix (ex: a notebook fix) you can not do it !

Even if the instance can be kept alive, you have no SSH access to it, so if you need to test some config change (ex: in `env.local` or in `docker-compose-extra.yml` or in `default.env`) you can not easily test it unless you spin up a new instance again ... which will be slow.

And finally once you are done with your testing and accumulated a bunch of running PAVICS instances I think there is no way to shut them down than to send an email to Francis to ask him to shut them down manually for you.

**Comment by Zeitsperre** [comment]
@tlvu @fmigneault @tlogan2000

I've tested the changes on an Ouranos infra service and have updated the associated notebooks in https://github.com/Ouranosinc/pavics-sdi/pull/328. Everything is passing for `pavics-sdi` and for `finch`.

If that's enough to satisfy people, I think we're safe to go forward with merging that PR as well as this one.

**Comment by Zeitsperre** [comment]
@fmigneault Just an FYI, but given the age of this PR, I figured it might be worthwhile to update `finch` further. I'm currently running local tests against the latest version (v0.13.2: June 2025) and if things are easy to update, it might be worthwhile to jump ahead.

Would we want to merge this for now and then propose the next update again in a separate PR?

**Comment by Zeitsperre** [comment]
@fmigneault The tests need to run against `pavics-sdi:update-for-finch-version-0.11.3`. Not sure what's going on with `raven-main/docs/source/notebooks/Subset_climate_data_over_watershed.ipynb`.

**Comment by Zeitsperre** [comment]
OK, I've gotten to the bottom of the issues with the failing cell in `raven-main/docs/source/notebooks/Subset_climate_data_over_watershed.ipynb`. It isn't great news.

The `clisops` installed in that image relies on `xemsf` in order to run `average_polygon`. Unfortunately, at around this time, that package was experiencing weird problems due to the need for an environment variable managed by `ESMF` that wasn't configured properly. The end result is that that specific process will always fail because `xesmf` can't load that environment variable.

In other words, `finch` v0.11.3 (and possibly several other subsequent versions) is borked. I'm testing later versions now to see if the issues still arise in later builds, but it isn't clear when exactly that `xesmf`/`ESMF` issue was fixed. All stop on this PR for now.

**Comment by Zeitsperre** [comment]
New build tests with `finch` v0.12.1:
- `pavics-sdi` notebooks are all passing (with https://github.com/Ouranosinc/pavics-sdi/pull/328)
- `raven` notebooks passing (on Ouranos and CRIM infra)
- Not certain why `RavenPy-main/docs/notebooks/paper/Perform_a_climate_change_impact_study_on_a_watershed.ipynb` is failing here. Passing on Ouranos infra without issue. Can't replicate the error locally.

@fmigneault and @tlvu Are we good to merge this and https://github.com/Ouranosinc/pavics-sdi/pull/328?

**Comment by Zeitsperre** [comment]
@fmigneault I need your review approval on this. Thanks!

**Comment by Zeitsperre** [comment]
I believe the tests are failing because in http://daccs-jenkins.crim.ca/job/PAVICS-e2e-workflow-tests/job/master/375/ because the finch being used is:
```
+ export FINCH_VERSION=0.9.2
```

And in http://daccs-jenkins.crim.ca/job/PAVICS-e2e-workflow-tests/job/master/374/ it isn't hitting the right `pavics-sdi` branch.

**Review by mishaschwartz — APPROVED** [review]
LGTM. As long as everyone else is happy with the update I have nothing else to add

**Comment by fmigneault** [comment]
The list of processes that https://github.com/Ouranosinc/pavics-sdi/blob/master/docs/source/notebook-components/weaver_example.ipynb indicates must also be updated in https://github.com/Ouranosinc/pavics-sdi/pull/328, since it relies on what `finch` WPS returns.

**Comment by Zeitsperre** [comment]
@fmigneault 

OK so this is really strange. Updated the outputs for `subset-user-input.ipynb` (see: https://github.com/Ouranosinc/pavics-sdi/commit/3de0cfe2956d6b20930b2ec1a6ff07e903682789) and re-ran on `kiwi` with the following configurations:
- `finch`: 0.12.1
- `pavics-sdi`: update-for-finch-version-0.11.3

Then I get this in my build logs:
```
13:52:28  =================================== FAILURES ===================================
13:52:28  _ pavics-sdi-update-for-finch-version-0.11.3/docs/source/notebooks/subset-user-input.ipynb::Cell 18 _
13:52:28  Notebook cell execution failed
13:52:28  Cell 18: Cell outputs differ
13:52:28  
13:52:28  Input:
13:52:28  help(finch.heat_wave_frequency)
13:52:28  
13:52:28  Traceback:
13:52:28   mismatch 'stdout'
13:52:28  
13:52:28   assert reference_output == test_output failed:
13:52:28  
13:52:28    'Help on meth...ut files.\n\n' == 'Help on meth...ut files.\n\n'
13:52:28    
13:52:28    Skipping 1404 identical leading characters in diff, use -v to show
13:52:28    Skipping 1694 identical trailing characters in diff, use -v to show
13:52:28    -    op : {'>=', '>', 'ge', 'gt'}string
13:52:28    ?           ^^               ^^
13:52:28    +    op : {'gt', '>', 'ge', '>='}string
13:52:28    ?           ^^               ^^
13:52:28  
```
Are you absolutely certain that the CRIM instance is running the right `finch` version?

For the `weaver_example.ipynb`, I keep getting `JSONDecodeErrors` when running on `kiwi`. I get the impression that the reason why we don't run the notebook by default is that it's not configured for Ouranos infra. Which environment variables specific to `weaver` are missing in the Jenkinsfile used in the `pavics-e2e-workflow-tests` repo?

Would it be possible for you to modify the CRIM Jenkinsfile to save this specific notebook as a run artifact?

**Comment by fmigneault** [comment]
@Zeitsperre 

Yes. 

The logs of http://daccs-jenkins.crim.ca/job/DACCS-iac-birdhouse/3577 that create the instance display: 

```shell
BIRDHOUSE_DEPLOY_BRANCH=update_finch_version
BUILD_DISPLAY_NAME=#3577
BUILD_ID=3577

# ...

+ grep -qE ^\s*FINCH_IMAGE\s*$
+ eval echo $FINCH_IMAGE
+ echo ${FINCH_DOCKER}:version-${FINCH_VERSION}
+ v=${FINCH_DOCKER}:version-${FINCH_VERSION}
+ eval echo "${FINCH_DOCKER}:version-${FINCH_VERSION}"
+ echo birdhouse/finch:version-0.12.1
+ value=birdhouse/finch:version-0.12.1
+ eval export ${i}="${value}"
+ export FINCH_IMAGE=birdhouse/finch:version-0.12.1
+ env
+ grep -e ^FINCH_IMAGE=
+ log DEBUG delayed eval 'FINCH_IMAGE=birdhouse/finch:version-0.12.1'
+ level=DEBUG
+ shift
+ [ delayed eval 'FINCH_IMAGE=birdhouse/finch:version-0.12.1' =  ]
``` 

The test run http://daccs-jenkins.crim.ca/job/PAVICS-e2e-workflow-tests/job/master/391 that is triggered with that instance then displays:

```shell
14:47:20  + echo TEST_PAVICS_SDI_REPO has been set to 'true'
14:47:20  TEST_PAVICS_SDI_REPO has been set to 'true'
14:47:20  + [ -z update-for-finch-version-0.11.3 ]
14:47:20  + echo PAVICS_SDI_BRANCH has been set to 'update-for-finch-version-0.11.3'
14:47:20  PAVICS_SDI_BRANCH has been set to 'update-for-finch-version-0.11.3'
14:47:20  + [ -z Ouranosinc/pavics-sdi ]
14:47:20  + echo PAVICS_SDI_REPO has been set to 'Ouranosinc/pavics-sdi'
14:47:20  PAVICS_SDI_REPO has been set to 'Ouranosinc/pavics-sdi'
```

> not configured for Ouranos infra

Indeed, the weaver component has to be enabled in the platform for it to work. 
The tests must enable both `TEST_PAVICS_SDI` and `TEST_PAVICS_SDI_WEAVER`.

**Comment by fmigneault** [comment]
I've re-triggered the workflow with `SAVE_RESULTING_NOTEBOOK=true`. 
http://daccs-jenkins.crim.ca/job/DACCS-iac-birdhouse/3583/

Just have to wait for them to complete and I can posted them here (if they don't timeout).

**Comment by fmigneault** [comment]
This time, the output is yet again another combination
```python
      "    op : {'ge', '>', 'gt', '>='}string\n",
      "        Comparison operation. Default: \">\".\n",
```

Could it be that a `set()` is used for that parameter definition, so the order is never guaranteed?

**Comment by tlvu** [comment]
Weaver nb fixed on CRIM pipeline (https://daccs-jenkins.crim.ca/job/PAVICS-e2e-workflow-tests/job/master/397/console).

We ignore Weaver nb failure on Ouranos because we do not run it and it's impossible to have the same output since the Weaver on Ouranos side detected more services (catalog, flyingpigeon, maleefowl) so the output will never match.

**Comment by tlvu** [comment]
run tests

To see the Weaver nb passed.

**Comment by Zeitsperre** [comment]
Something isn't adding up here.

Ouranos:

```
00:55:50  =================================== FAILURES ===================================
00:55:50  _ pavics-sdi-update-for-finch-version-0.11.3/docs/source/notebooks/subset-user-input.ipynb::Cell 18 _
00:55:50  Notebook cell execution failed
00:55:50  Cell 18: Cell outputs differ
00:55:50  
00:55:50  Input:
00:55:50  help(finch.heat_wave_frequency)
00:55:50  
00:55:50  Traceback:
00:55:50   mismatch 'stdout'
00:55:50  
00:55:50   assert reference_output == test_output failed:
00:55:50  
00:55:50    'Help on meth...ut files.\n\n' == 'Help on meth...ut files.\n\n'
00:55:50    
00:55:50    Skipping 1404 identical leading characters in diff, use -v to show
00:55:50    Skipping 1694 identical trailing characters in diff, use -v to show
00:55:50    -    op : {'>=', '>', 'ge', 'gt'}string
00:55:50    ?           ^^               ^^
00:55:50    +    op : {'gt', '>', 'ge', '>='}string
00:55:50    ?        
```

CRIM:
```
01:17:55  =================================== FAILURES ===================================
01:17:55  _ pavics-sdi-update-for-finch-version-0.11.3/docs/source/notebooks/subset-user-input.ipynb::Cell 18 _
01:17:55  Notebook cell execution failed
01:17:55  Cell 18: Cell outputs differ
01:17:55  
01:17:55  Input:
01:17:55  help(finch.heat_wave_frequency)
01:17:55  
01:17:55  Traceback:
01:17:55   mismatch 'stdout'
01:17:55  
01:17:55   assert reference_output == test_output failed:
01:17:55  
01:17:55    'Help on meth...ut files.\n\n' == 'Help on meth...ut files.\n\n'
01:17:55    
01:17:55    Skipping 1411 identical leading characters in diff, use -v to show
01:17:55    Skipping 1694 identical trailing characters in diff, use -v to show
01:17:55    -  {'gt', '>=', '>', 'ge'}string
01:17:55    ?         ------
01:17:55    +  {'gt', '>', 'ge', '>='}string
01:17:55    ?                 ++++++
01:17:55  
```

Is it possible that the docstring is being rendered as a Python object (as `sets` are unordered)?

**Comment by tlvu** [comment]
> @tlvu I've added an `# NBVAL_IGNORE_OUTPUT` on that particular cell. We should be good to move forward with this now.

Isn't this a simple fix to order the set before output?  Something like `sorted(set.values())` ?

`# NBVAL_IGNORE_OUTPUT` is when we don't have a choice because the format will always change.  Here we change the code once and it's good for all time.

**Comment by Zeitsperre** [comment]
> Isn't this a simple fix to order the set before output? Something like `sorted(set.values())` ?

It isn't so simple; The docstring for the object is being created via the `birdy.WPSClient` class. The docstring that is parsed and presented via the `help` built-in is a string:

```
"Number of heat waves over a given period. A heat wave is defined as an event where the minimum and maximum daily temperature both exceeds specific thresholds over a minimum number of days.\n\nParameters\n----------\ntasmin : ComplexData:mimetype:`application/x-netcdf`, :mimetype:`application/x-ogc-dods`\n (etc.)"
```
We would have to modify the docstring-parsing mechanism in `birdy` in order to do this (which would then involve a new release of `birdy`, testing, integration, etc.).

Otherwise, we could add a cell that performs a cleanup of the `finch.heat_wave_frequency.__doc__`, but that would involve coming up with a regex string that looks for `{('>'|'>='|'gt'|'ge')}` and replaces it with an entry with a specific order. It just doesn't seem worth the trouble.

> `# NBVAL_IGNORE_OUTPUT` is when we don't have a choice because the format will always change. Here we change the code once and it's good for all time.

We could look at the mechanism in `birdy`. I can certainly open an issue about it. If we want to actually address this problem (as it will definitely become more pertinent with newer versions of `xclim` that have indices using the `op` argument), we could fix it there.

**Comment by tlvu** [comment]
> We could look at the mechanism in `birdy`. I can certainly open an issue about it.

Oh boy the fix is actually `birdy` ?!  Ok open the issue on birdy's side to not forget about this, then let's use `# NBVAL_IGNORE_OUTPUT` to finish with this PR.

Note the new birdy issue here so we have a trace.

**Comment by tlvu** [comment]
> Note the new birdy issue here so we have a trace.

Reference the issue https://github.com/bird-house/birdy/issues/277 in this PR as well so we have a complete cross-reference.

**Comment by Zeitsperre** [comment]
See: https://github.com/bird-house/birdhouse-deploy/pull/354#issue-1803575832

**Comment by fmigneault** [comment]
> Ok open the issue on birdy's side to not forget about this, then let's use `# NBVAL_IGNORE_OUTPUT` to finish with this PR.

I would rather have `birdy` updated and integrated. Using `# NBVAL_IGNORE_OUTPUT` defeats the purpose of the test.

**Comment by fmigneault** [comment]
http://daccs-jenkins.crim.ca/job/DACCS-iac-birdhouse/3596/

**Review by fmigneault — APPROVED** [review]
OK for finch to avoid delay of birdy release impacting docker birdhouse/pavics switch.
